### PR TITLE
feat: add GeoVex embedding model

### PR DESCRIPTION
**Pull request by mschrader15**
[GeoVex](https://openreview.net/forum?id=7bvWopYY1H) is a geo-spatial embedding model proposed by @ddonghi & @annemorvan 

We (http://greenlastmile.ai/) have found that it improves on the performance of Hex2Vec in regression & clustering tasks.

I have implemented their model in the `srai` library and provided an example qualitative comparison notebook (albeit using a smaller hexagonal neighborhood and tag set).

### CODE REVIEW

**Comment by piotrgramacki** [comment]
Hi, @mschrader15 
Thank you for your contribution! We appreciate the time and effort put into this work! This is a valuable addition to our library since we already had our eyes on GeoVex. We will review your PR this week and hopefully add it to `srai` soon. 

If you have any insights or suggestions after using our library, we would love to hear about them. Our library is still in the early stage of development, so we assume that there are many areas to improve ;) 

P.S. Are you aware of any potential copyright issues with adding GeoVex to `srai`?

**Comment by mschrader15** [comment]
Yea no problem @piotrgramacki! I am really excited about this library and hope to contribute further. We developed many of the same capabilities in parallel for our project, but haven't found the time to switch over to `srai` completely. I'll slowly get around to some future PRs.

As far as copyright goes, I am no expert, but I reached out to the GeoVex team about this PR and they were in support (cc. @ddonghi & @annemorvan). My implementation was based purely on the paper that was available on OpenReview

**Review by piotrgramacki — CHANGES_REQUESTED** [review]
Sorry for the delay, but I finally had time to go through your code. Overall looks nice, but I have two things:

1. I left some comments, which are worth addressing IMHO
2. Tests ;) We try to keep our library covered with tests, so if it's not a problem we would like to ask you to add them. I would start with tests for the dataset - there is some logic there, which we know how it should work. Then embedder - mainly to check if it works correctly with incorrect input, unfitted model, etc. There is no need to check embeddings; I would only verify if the format of returned values is correct. 

If you have any questions feel free to ask ;)

File: srai/embedders/geovex/embedder.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+        # save invalid h3s for later
+        self._invalid_h3s: List[str] = []
+
+    def transform(
+        self,
+        regions_gdf: gpd.GeoDataFrame,
+        features_gdf: gpd.GeoDataFrame,
+        joint_gdf: gpd.GeoDataFrame,
+    ) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (mschrader15) [reply srai/embedders/geovex/embedder.py:86]: Good observation. Will fix. 

FWIW, as an outsider who has only used this library to implement the GeoVex example, the need for 3 dataframes + a neighborhood object to describe a region feels a bit excessive? Seems like there is potentially a more concise way to represent an area?

[the remark this answers is not in the document]
Comment (piotrgramacki) [reply srai/embedders/geovex/embedder.py:86]: Well, you're right on that. We have actually already drafted a proposal to replace it with a single object, but it is WIP. Maybe @simonusher can tell a bit more about how we plan to approach this

[the remark this answers is not in the document]
Comment (mschrader15) [reply srai/embedders/geovex/embedder.py:91]: Awesome! Also mypy through an error when I added the neighborhood as an argument to this function. Instead I just added the re-creation of the neighborhood object inside of the function.

File: examples/embedders/geovex_embedder.ipynb
[the remark this answers is not in the document]
Comment (mschrader15) [reply examples/embedders/geovex_embedder.ipynb:1]: I agree it would be nice to be a method somewhere, but it needs to happen prior to the download of data. 

Do you think the function belongs in GeoVex module or somewhere else?

[the remark this answers is not in the document]
Comment (RaczeQ) [reply examples/embedders/geovex_embedder.ipynb:1]: I guess we could put this function `new_geometry` in `srai.h3` module. I can see it looking like this:

```python
from srai.h3 import h3_to_geoseries, buffer_h3_cells  # buffer_h3_cells being the new function
resolution = 9
k_ring_buffer_radius = 4
regionalizer = H3Regionalizer(resolution=resolution)

cells = regionalizer.transform(geometry)  # get all the cells for given region / geometry
buffered_cells = buffer_h3_cells(cells=cells, k_ring_buffer_radius=k_ring_buffer_radius)  # buffer them using this new function
new_geometry = h3_to_geoseries(buffered_cells).unary_union  # get all cells geometries and union them to a single geometry
```

In this `buffer_h3_cells` function we can utilize `from h3ronpy.pandas import grid_disk` function (https://h3ronpy.readthedocs.io/en/latest/usage/grid.html#grid-disks-with-h3ronpy-pandas-grid-disk).

File: srai/neighbourhoods/h3_neighbourhood.py [hunk 2]
```diff
@@ -87,6 +87,20 @@ def get_neighbours_at_distance(self, index: str, distance: int) -> Set[str]:
         neighbours.discard(index)
         return self._select_available(neighbours)
 
+    def get_ij_index(self, anchor_index: str, target_index: str) -> Tuple[int, int]:
```
[the remark this answers is not in the document]
Comment (RaczeQ) [reply srai/neighbourhoods/h3_neighbourhood.py]: I have prepared a dedicated function `get_local_ij_index` in the `srai.h3` module. You can just substitute it with `get_local_ij_index(anchor_index, target_index)`. But it returns normalized distance, always bound to the anchor (it subtracts the origin index, to make it 0,0 in the center), and I don't know if it will break anything in the GeoVex methodology.

File: examples/embedders/geovex_embedder.ipynb
[the remark this answers is not in the document]
Comment (mschrader15) [reply examples/embedders/geovex_embedder.ipynb:1]: I added the function in [e29b0d4](https://github.com/kraina-ai/srai/pull/293/commits/e29b0d48920ef17875dc0dee7ea68a3a14829bf4)

File: srai/h3.py [hunk 3]
```diff
@@ -173,3 +174,42 @@ def get_local_ij_index(
         local_ijs = [(coords[0], coords[1]) for coords in local_ijs]
 
     return local_ijs
+
+
+def ring_buffer_geometry(
```
Comment (mschrader15) [inline srai/h3.py:179]: Ref https://github.com/kraina-ai/srai/pull/293#discussion_r1330301257

**Comment by mschrader15** [comment]
Glad to finally find time to work on this PR again. 

I'll slowly add further test cases in the coming days...

**Comment by RaczeQ** [comment]
Hi @mschrader15, that's almost exactly what I had in mind with the `ring_buffer_geometry`. I've looked deeper into the example notebook and tried to refine the flow of it.
As a result, I made some changes to the main codebase of the library and altered your implementation of the new method in the `srai.h3` module. I also changed the example notebook to use those new functions.

I've created a PR on your fork of the library directly to this branch: https://github.com/mschrader15/srai/pull/1
If you have any feedback about those changes, you're welcome to share it 😄
By the way, great job on implementing the model. The difference between GeoVex and hex2vec is striking, and the RGB visualizations are beautiful.

And sorry for the long delay after the last comment.

File: tests/embedders/geovex/test_embedder.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    """Test that Hex2VecEmbedder raises ModelNotFitException if not fitted."""
+    embedder = GeoVexEmbedder(
+        [f"{t}_{st}" for t in HEX2VEC_FILTER for st in HEX2VEC_FILTER[t]],  # type: ignore
+        H3Neighbourhood(),
+    )
+    with pytest.raises(ModelNotFitException):
+        embedder.transform(gpd.GeoDataFrame(geometry=[]), gpd.GeoDataFrame(), gpd.GeoDataFrame())
+
+
+def test_embedder() -> None:
```
Comment (mschrader15) [inline tests/embedders/geovex/test_embedder.py:51]: Missing the files for this. The pre-commit hook `check-added-large-files` prevented adding them. Can't get the `*_features.parquet` small enough with the >= 256 features requirement.

**Comment by mschrader15** [comment]
Hey @RaczeQ, I just added test coverage for GeoVex.

I kept your PR to my branch as-is, all looked good to me. 

I did switch the example notebook from London to Wroclaw, as London's file size & training time didn't seem appropriate for an example. 

See the note about the embedder `fit_transform` test. As of now I have a hard limit on the minimum # of features being >= 256 (in channels to first conv. layer) to keep with the author's implementation.

**Comment by mschrader15** [comment]
I found an error in the hexagonal weights of the loss function. 

Wait to close this PR until I push a fix 🤦‍♂️

File: srai/embedders/geovex/embedder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonusher) [reply srai/embedders/geovex/embedder.py:93]: Hey, I'm very sorry for not responding to this earlier. I agree with what you said -> the API is a bit cumbersome and we will be working to improve it. We overlooked some use cases when designing it, then added them afterwards. At first, we also thought this would be more of an internal API and that we would write an easier one on top of it (which will probably happen anyway).
The way we approach this will probably also be based on the results we have with using dask for out-of-core processing and trying out duckdb.
What I've been playing with was just having a container object like [HF's BatchEncoding](https://huggingface.co/docs/transformers/v4.34.0/en/main_classes/tokenizer#transformers.BatchEncoding) but I'm not sure if we'll go this exact way. Anyway, this requires some more design work. Feel free to share any ideas if you'd like :)

**Comment by mschrader15** [comment]
[0e91bbd](https://github.com/kraina-ai/srai/pull/293/commits/0e91bbd6c76ba62e586a53f454d79507b3b0e659) fixes https://github.com/kraina-ai/srai/pull/293#issuecomment-1763250613

File: srai/embedders/geovex/embedder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mschrader15) [reply srai/embedders/geovex/embedder.py:93]: >  We overlooked some use cases when designing it, then added them afterwards. At first, we also thought this would be more of an internal API and that we would write an easier one on top of it (which will probably happen anyway).

Totally understand that! 

Re dask, when we implemented your Hex2Vec model, we were training on city metro-regions, and had to switch to dask out-of-core for the spatial join of the larger regions and had success. 

A database like DuckDB could be the best solution, where the relationship between regions, features, joint, and ultimately embeddings happens behind the scenes so to say. Then again ig there is a fine line to walk between accessibility, ease-of-use, and performance

File: tests/embedders/geovex/test_embedder.py (hunk 4 above)
Comment (RaczeQ) [reply tests/embedders/geovex/test_embedder.py:51]: Do you know how big was this file? We have increased it to 1mb from the default 500kb for another use case.

Comment (mschrader15) [reply tests/embedders/geovex/test_embedder.py:51]: The features file is 22MB...

I will make the conv in-channels a parameter (which it should be anyways), then fitting inside the 1MB limit will be do-able

Comment (mschrader15) [reply tests/embedders/geovex/test_embedder.py:53]: [29bd6cf](https://github.com/kraina-ai/srai/pull/293/commits/29bd6cff08cade7b1e08c33d0bd7064cac778370) fits into the 1MB requirement

File: tests/embedders/geovex/generation.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
+"""Test case generation for Hex2VecEmbedder."""
```
Comment (simonusher) [inline tests/embedders/geovex/generation.py:1]:
```suggestion
"""Test case generation for GeoVexEmbedder."""
```

File: tests/embedders/geovex/generation.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    if tags is None:
+        tags = HEX2VEC_FILTER
+
+    neighbourhood = H3Neighbourhood()
+    regions_indexes = neighbourhood.get_neighbours_up_to_distance(
+        root_region_index, region_gen_radius
+    )
+    regions_indexes.add(root_region_index)
+    regions_indexes = list(regions_indexes)  # type: ignore
```
Comment (simonusher) [inline tests/embedders/geovex/generation.py:43]: NIT: Same as somewhere below - you could maybe use `include_center` here

File: tests/embedders/geovex/constants.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+"""Constants for hex2vec tests."""
```
Comment (simonusher) [inline tests/embedders/geovex/constants.py:1]:
```suggestion
"""Constants for geovex tests."""
```

File: srai/embedders/geovex/embedder.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+try:  # pragma: no cover
+    from torch.utils.data import DataLoader
+
+except ImportError:
+    from srai.embedders._pytorch_stubs import DataLoader
+
+
+class GeoVexEmbedder(CountEmbedder):
+    """Hex2Vec Embedder."""
```
Comment (simonusher) [inline srai/embedders/geovex/embedder.py:33]:
```suggestion
    """GeoVex Embedder."""
```

**Review by simonusher — CHANGES_REQUESTED** [review]
Great job! Once again, thank you for this contribution.
I only found some minor issues, like comments left after copying a template from other files (Hex2Vec instead of GeoVex in some places). So, I would only ask you to address that. Other than that it looks very good

**Comment by mschrader15** [comment]
No problem! Good catch on these comments/variable names. 

Took at stab at correcting all the issues you pointed out. Also gave credit to the orginal authors in the change log.

File: srai/embedders/geovex/embedder.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+        )
+
+        self._model: Optional[GeoVexModel] = None
+        self._is_fitted = False
+        self._r = neighbourhood_radius
+        self._embedding_size = embedding_size
+        self._convolutional_layers = convolutional_layers
+        self._convolutional_layer_size = convolutional_layer_size
+
+        self._neighbourhood = neighbourhood
```
Comment (RaczeQ) [inline srai/embedders/geovex/embedder.py:78]: Attribute is not used in the embedder.

File: srai/embedders/geovex/embedder.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        self, regions_gdf: pd.DataFrame, features_gdf: pd.DataFrame, joint_gdf: pd.DataFrame
+    ) -> pd.DataFrame:
+        return super().transform(regions_gdf, features_gdf, joint_gdf).astype(np.float32)
+
+    def _check_is_fitted(self) -> None:
+        if not self._is_fitted or self._model is None:
+            raise ModelNotFitException("Model not fitted. Call fit() or fit_transform() first.")
+
+    def _prepare_trainer_kwargs(self, trainer_kwargs: Optional[Dict[str, Any]]) -> Dict[str, Any]:
+        # TODO: this is copy pasted from Hex2VecEmbedder, should be refactored
```
Comment (RaczeQ) [inline srai/embedders/geovex/embedder.py:247]: Comment to remove or to implement?

File: srai/embedders/geovex/dataset.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+    def _assert_h3_neighbourhood(self, neighbourhood: H3Neighbourhood) -> None:
+        # force that the neighbourhood is an H3Neighbourhood,
+        # because we need the get_ij_index method
+        if not isinstance(neighbourhood, H3Neighbourhood):
+            raise ValueError(
+                f"neighbourhood has to be an H3Neighbourhood, but was {type(neighbourhood)}"
+            )
+
+    def get_ordered_index(self) -> List[str]:
```
Comment (RaczeQ) [inline srai/embedders/geovex/dataset.py:180]: Should it be named `get_ordered_index` or `get_valid_cells` (looking at the function below)? I see that there is no sorting / ordering in this function.

Comment (mschrader15) [reply srai/embedders/geovex/dataset.py:180]: It is sorted per-say, as it maintains the order of the input count Dataframe, but I am okay with the rename

**Comment by RaczeQ** [comment]
I'm trying to pinpoint the issue of the `tests/embedders/geovex/test_embedder.py::test_embedder` failures.
All I can see is that only Python 3.8 doesn't work on Github Action runner, even though it works locally for me 🙃 
I thought it was a neighbourhood issue since the order of neighbours is deterministic in sets, but the order is different in every run and still versions 3.9-3.11 work flawlessly.

**Comment by mschrader15** [comment]
Works for me locally on 3.9 too. Potentially something deterministic with [Pytorch](https://pytorch.org/docs/stable/notes/randomness.html#avoiding-nondeterministic-algorithms)?

File: srai/embedders/geovex/embedder.py [hunk 12]
```diff
@@ -163,7 +163,13 @@ def fit(
 
         trainer_kwargs = self._prepare_trainer_kwargs(trainer_kwargs)
         counts_df, dataloader, dataset = self._prepare_dataset(  # type: ignore
-            regions_gdf, features_gdf, joint_gdf, neighbourhood, self._batch_size, shuffle=True
+            regions_gdf,
+            features_gdf,
+            joint_gdf,
+            neighbourhood,
+            self._batch_size,
+            shuffle=False,
```
Comment (mschrader15) [inline srai/embedders/geovex/embedder.py:171]: I would contend that we we should leave `shuffling=True` by default.

**Comment by RaczeQ** [comment]
@mschrader15 I've managed to fix the issue with reproducibility.

It turns out that weights and biases initialization wasn't deterministic (probably with additional modules required for GeoVex).
To fix it I have added parameters initialization with override in tests to make it consistent.

**Comment by mschrader15** [comment]
> * change the accelerators from CPU to AUTO in the notebook

The [torch.lgamma function does not work on Apple MPS](https://github.com/pytorch/pytorch/issues?q=lgamma+mps), which torch defaults to if available.

**Review by piotrgramacki — APPROVED** [review]

**Comment by RaczeQ** [comment]
After a long time, we have finally managed to check everything and merge the GeoVex model into the library 🥳.

Thank you @mschrader15 for this huge contribution and we're sorry that it took so long. We will create a new release soon to make it available to everyone. 

I hope that future contributions won't take so long to review 😅.

**Comment by mschrader15** [comment]
@RaczeQ no problem, I sat on it for a couple months myself! Thanks for all the help and detailed reviews! In the end its a really powerful tool that I'm glad is out in the OS community. (p.s. excited to see what comes from #352 )

Yep I'm totally okay with a Social Media post. Maybe just mention that I did it as part of "the Green Last Mile group, supported by a [Climate Change AI] (https://www.linkedin.com/company/climatechangeai/) grant"

**Comment by ddonghi** [comment]
Thanks everyone for having released GeoVex to the OS community! Love to see it! 👍
